substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs

```diff
@@ -552,37 +554,34 @@ where
 			// The best reported block is in the pruned list. Report a new best block.
 			let is_in_pruned_list =
 				pruned_block_hashes.iter().any(|hash| *hash == current_best_block);
-			// The block is not the last finalized block.
+
+			// The best reported block is not the last finalized block.
 			//
 			// It can be either:
 			//  - a descendant of the last finalized block
 			//  - a block on a fork that will be pruned in the future.
 			//
 			// In those cases, we emit a new best block.
+			//
+			// When [`Self::generate_finalized_events`] reports the last finalized block as the best
+			// block, it also sets the `self.current_best_block`, if and only if the block
+			// was never reported as `NewBlock` before.
+			//
+			// Therefore, this condition is equivalent with
+			// `!std::matches(FollowEvent::BestBlockChanged(_), events.last())`.
+			// And that means we know for sure that at least a `NewBlock` was generated
+			// for the last finalized block.
```

Comment:
How do we know? All of the blocks could already have been known. Then there will be no `NewBlock` generated?

Reply:
Have removed this comment and added a simpler code path using lowest common ancestor, thanks!